### PR TITLE
Dashboard: refactor enzyme tests to use react testing library

### LINKED ISSUE #163: Migrate `Enzyme` tests to `React Testing Library` tests
All tests within the `Dashboard` package are using `Enzyme`. Instead, the tests inside the `react-component` package are using the` React Testing Library` (RTL). To use only one Test Library, it was decided to migrate the tests from `Enzyme` to `RTL`.

### PR DESCRIPTION
**Pull request by vallq**
## What's new

fixes #163 

- [x] app
- [x] app-bar
- [x] dashboard
- [x] emergency-alarm
- [x] negotiation panel //unable to refactor the first test (not sure if it is necessary as well)
- [x] main menu

**Schedule-visualizer**
- [x] index
- [x] dispensers-overlay
- [x] doors-overlay
- [x] lift-overlay
- [x] trajectories-overlay
- [x] robots-overlay
- [x] waypoints-overlay

**Auth**
- [x] login //see comment in file
- [x] private route //last test unable to pass

**Drawers**
- [x] drawer-header
- [x] help-drawer
- [x] hotkeys-dialog
- [x] settings-drawer

Error pages
- [x] page not found
- [x] unauthorized

## Self-checks

- [x] I'm familiar with and follow this [Typescript guideline](https://basarat.gitbook.io/typescript/styleguide)
- [ ] I added unit-tests for new components
- [ ] I tried testing edge cases
- [ ] I tested the behaviour of the components that interact with the backend, with an e2e test

## Discussion

The tests that I have not been able to resolve are
- Negotiation Panel: first test, I am not sure if it is actually needed
- Login: the third test is not needed but without it, the last two tests will always fail. but when the tests are run individually, they will pass
- Private Route: unable to get the last test to pass

### CODE REVIEW

**Comment by vallq** [comment]
Hi @koonpeng @matiasbavera @ChawinTan 
I've updated most of the tests with a few that I have not been able to resolve, so if you have any idea please do let me know!
feel free to comment and give feedback as well!

File: packages/dashboard/src/components/dashboard/tests/negotiations-panel.test.tsx [hunk 1]
```diff
@@ -154,14 +151,16 @@ it('should empty all current negotiations when clear button is clicked', () => {
       setNegotiationTrajStore={setNegotiationTrajStore}
     />,
   );
-  root.find('button#clear-button').simulate('click');
-
-  expect(root.find('li').length).toEqual(0);
+  const clear = root.getByRole('button', {
+    name: /Clear/i,
+  });
+  fireEvent.click(clear);
```
Comment (matiasbavera) [inline packages/dashboard/src/components/dashboard/tests/negotiations-panel.test.tsx:157]: same comment as above

File: packages/dashboard/src/components/dashboard/tests/negotiations-panel.test.tsx [hunk 2]
```diff
@@ -1,16 +1,15 @@
-import { createMount, createShallow } from '@material-ui/core/test-utils';
+import { createShallow } from '@material-ui/core/test-utils';
```
Comment (matiasbavera) [inline packages/dashboard/src/components/dashboard/tests/negotiations-panel.test.tsx:1]: Not sure if this is the correct import

File: packages/dashboard/src/components/tests/app.test.tsx [hunk 3]
```diff
@@ -1,15 +1,14 @@
...
-const mount = createMount();
-
 // react-leaflet doesn't work well in jsdom.
 jest.mock('./../schedule-visualizer', () => () => null);
 
-it('renders correctly', async () => {
-  await act(async () => {
-    expect(mount(<App />)).toMatchSnapshot();
+it('renders without crashing', async () => {
+  let root: any;
```
Comment (matiasbavera) [inline packages/dashboard/src/components/tests/app.test.tsx:9]: nit: You can replace the any for the inferred type

File: packages/dashboard/src/components/schedule-visualizer/tests/waypoints-overlay.test.tsx [hunk 4]
```diff
@@ -13,14 +12,21 @@ test('Render waypoints correctly', async () => {
...
+      ]}
+    >
       <WaypointsOverlay bounds={bounds} currentLevel={currentLevel} />
     </LMap>,
   );
 
-  expect(wrapper.find(WaypointMarker).exists()).toBeTruthy();
-  expect(wrapper.find(WaypointMarker).length).toBe(waypoints.length);
+  await waitFor(() => {
+    expect(root.queryAllByTestId('waypointMarker')).toBeTruthy();
```
Comment (matiasbavera) [inline packages/dashboard/src/components/schedule-visualizer/tests/waypoints-overlay.test.tsx:27]: This a dangerous assertion because empty [] in some cases is true.

`if ([]) {console.log("It's true!")}; ` Run this in your browser's console.

So you may check .length > 0

HTH

File: packages/dashboard/src/components/dashboard/tests/negotiations-panel.test.tsx [hunk 5]
```diff
@@ -124,8 +123,8 @@ it('renders negotiations correctly', () => {
   root.unmount();
 });
 
-it('matches snapshot', () => {
-  const root = mount(
+it('renders without crashing', () => {
```
Comment (matiasbavera) [inline packages/dashboard/src/components/dashboard/tests/negotiations-panel.test.tsx:126]: Smoke test?

File: packages/dashboard/src/components/dashboard/tests/negotiations-panel.test.tsx (hunk 1 above)
Comment (matiasbavera) [inline packages/dashboard/src/components/dashboard/tests/negotiations-panel.test.tsx:157]: Maybe, userEvent?

File: packages/dashboard/src/components/drawers/tests/drawer-header.test.tsx [hunk 6]
```diff
@@ -1,49 +1,44 @@
...
+    const root = render(<DrawerHeader handleCloseButton={() => jest.fn()} title={'Test'} />);
+    expect(root.queryByTestId('backDrawerButton')).toBeFalsy();
   });
 
   test('Click on close button fires an event correctly', () => {
     let clicked = false;
-    const root = mount(<DrawerHeader handleCloseButton={() => (clicked = true)} title={'Test'} />);
-    root.find('#closeDrawerButton').simulate('click');
+    const root = render(<DrawerHeader handleCloseButton={() => (clicked = true)} title={'Test'} />);
+    root.getByTestId('closeDrawerButton').click();
```
Comment (matiasbavera) [inline packages/dashboard/src/components/drawers/tests/drawer-header.test.tsx:28]: I'm curious, Is this working? with the click in the end?

File: packages/dashboard/src/components/error-pages/tests/page-not-found.test.tsx [hunk 7]
```diff
@@ -1,15 +1,15 @@
-import { mount } from 'enzyme';
 import React from 'react';
+import { render } from '@testing-library/react';
 import NotFoundPage from '../page-not-found';
 import { BrowserRouter } from 'react-router-dom';
 
 describe('PageNotFound', () => {
   test('renders correctly', () => {
```
Comment (matiasbavera) [inline packages/dashboard/src/components/error-pages/tests/page-not-found.test.tsx:7]: smoke test? Since we are not comparing against a snapshot anymore

**Review by matiasbavera — CHANGES_REQUESTED** [review]
Hi Val, thanks for this, it's looking good, left a couple of comments.

Here are some best practices and common mistakes when writing test with RTL (Written by the author of  RTL) https://kentcdodds.com/blog/common-mistakes-with-react-testing-library/ 
HTH

File: packages/dashboard/src/components/auth/tests/login.test.tsx [hunk 8]
```diff
@@ -18,6 +19,17 @@ describe('Login page', () => {
...
+      </BrowserRouter>,
+    );
+    root.unmount();
+  });
+
+  /*NOTE: Removing this test causes the subsequent 2 tests to fail
+  but running the tests individually (test.only) allows the tests to pass*/
   test('redirects to dashboard when user is authenticated', () => {
     window.history.replaceState(window.history.state, '');
     const wrapper = mount(
```
[the remark this answers is not in the document]
Comment (vallq) [reply packages/dashboard/src/components/auth/tests/login.test.tsx:35]: yea, the third test is enzyme, everything else is RTL.
I didn't try with a cleanup, I can look into that today!

File: packages/dashboard/src/components/auth/tests/login.test.tsx [hunk 9]
```diff
@@ -34,20 +46,37 @@ describe('Login page', () => {
...
           <Login />
         </AuthenticatorContext.Provider>
       </BrowserRouter>,
     );
-    const loginButton = root.find('button#login-button').first();
-    expect(loginButton).toBeTruthy();
-    loginButton.simulate('click');
+    //screen.debug();
+    const loginButton = root.getByRole('button', { name: /Login with RMF/i });
+    fireEvent.click(loginButton);
```
[the remark this answers is not in the document]
Comment (vallq) [reply packages/dashboard/src/components/auth/tests/login.test.tsx:79]: Okay got it, I will change the relevant ones!

File: packages/dashboard/src/components/auth/tests/private-route.test.tsx [hunk 10]
```diff
@@ -59,4 +60,15 @@ describe('PrivateRoute', () => {
     expect(history.location.pathname).toBe('/private');
     expect(component.find(Unauthorized)).toBeTruthy();
   });
+
+  /* This test is unable to pass
```
[the remark this answers is not in the document]
Comment (vallq) [reply packages/dashboard/src/components/auth/tests/private-route.test.tsx:64]: I tried with waitFor, but it just causes the test to never end and I would have to manually stop it with ctrl+c
```
 RUNS  src/components/auth/tests/private-route.test.tsx
Test Suites: 0 of 1 total
Tests:       3 passed, 3 total
Snapshots:   0 total
Time:        17 s
```
will try out with act!

File: packages/dashboard/src/components/dashboard/tests/negotiations-panel.test.tsx (hunk 2 above)
Comment (vallq) [reply packages/dashboard/src/components/dashboard/tests/negotiations-panel.test.tsx:1]: yea this is used in the very first test so I have not been successful in replicating this test using RTL. and I'm not sure if it's actually necessary?

File: packages/dashboard/src/components/schedule-visualizer/tests/waypoints-overlay.test.tsx (hunk 4 above)
Comment (vallq) [reply packages/dashboard/src/components/schedule-visualizer/tests/waypoints-overlay.test.tsx:27]: ohh okay, I overlooked this point! thanks for highlighting

File: packages/dashboard/src/components/drawers/tests/drawer-header.test.tsx (hunk 6 above)
Comment (vallq) [reply packages/dashboard/src/components/drawers/tests/drawer-header.test.tsx:28]: yea it actually does but it uses the [web API](https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/click).. so I'll probably change this to userEvent instead, as per your previous comments?

putting a console log line before and after the clicks to log the variable 'clicked':
```
console.log
    false
      at Object.<anonymous> (src/components/drawers/tests/drawer-header.test.tsx:41:13)
  console.log
    true
      at Object.<anonymous> (src/components/drawers/tests/drawer-header.test.tsx:43:13)
```

**Comment by vallq** [comment]
Updates:
- Replaced 'fireEvent' with 'userEvent'
- Removed unused lines
- Fixed flakey tests

File: packages/dashboard/src/components/dashboard/tests/negotiations-panel.test.tsx [hunk 11]
```diff
@@ -74,58 +69,19 @@ it('renders negotiations correctly', () => {
...
 
-  root.unmount();
+  userEvent.click(finishedLabel);
+  const rejectedLabel = root.getAllByRole('treeitem', { name: /[REJECTED]/i });
+  expect(rejectedLabel.length).toBe(1);
 });
 
-it('matches snapshot', () => {
-  const root = mount(
+it('renders without crashing', () => {
```
Comment (matiasbavera) [inline packages/dashboard/src/components/dashboard/tests/negotiations-panel.test.tsx:83]: I think we can remove this test

File: packages/dashboard/src/setupTests.ts [hunk 12]
```diff
@@ -1,5 +1,6 @@
 import Enzyme from 'enzyme';
 import Adapter from 'enzyme-adapter-react-16';
```
Comment (matiasbavera) [inline packages/dashboard/src/setupTests.ts:2]: From this PR and on,  Enzyme will no longer be used, we can remove the enzyme dependencies from `package.json`. And from here too.

**Review by matiasbavera — COMMENTED** [review]
Thanks Val, added some comments

File: packages/dashboard/src/setupTests.ts (hunk 12 above)
Comment (vallq) [reply packages/dashboard/src/setupTests.ts:2]: got it!

Comment (vallq) [reply packages/dashboard/src/setupTests.ts:2]: hi @matiasbavera , for this current PR, is it alright to leave these import lines first? removing them will break the tests in the react tour component and I think it's more efficient to address that in a separate PR. there are several parts in the tour component that needs to be updated with the changes in the dashboard (eg. using a stack for the menu navigation)

**Comment by vallq** [comment]
Hi @matiasbavera I have removed the negotiation panel test!

File: packages/dashboard/src/setupTests.ts (hunk 12 above)
Comment (matiasbavera) [reply packages/dashboard/src/setupTests.ts:2]: Yeah, you're right!

File: packages/dashboard/src/components/schedule-visualizer/tests/robot-trajectories-overlay.test.tsx [hunk 13]
```diff
@@ -87,8 +55,23 @@ describe('RobotTrajectoriesOverlay', () => {
...
+            <RobotTrajectoriesOverlay
+              bounds={mapBound}
+              robots={robots}
+              trajectories={[trajectoryValue]}
+              conflicts={trajectoryConflict}
+            />
+          </SettingsContext.Provider>
+        </LMap>,
+      );
+      expect(root.queryAllByTestId('trajMarker')).toBeTruthy();
```
Comment (matiasbavera) [inline packages/dashboard/src/components/schedule-visualizer/tests/robot-trajectories-overlay.test.tsx:73]: Hi @vallq one last thing, this is the same thing that I explained before if([]) is `true`. So I added that in my version, and it hangs the test. So, that assertion it's not working. Can you try it on your pc? If it's not working for you too, maybe it is not the way of doing things. So can you dig a bit about it. If you can find a solution, we can use it as a smoke test (renders without an assertion).

**Review by matiasbavera — CHANGES_REQUESTED** [review]
Hi, Val left one relevant comment.

HTH,

File: packages/dashboard/src/components/schedule-visualizer/tests/robot-trajectories-overlay.test.tsx (hunk 13 above)
Comment (vallq) [reply packages/dashboard/src/components/schedule-visualizer/tests/robot-trajectories-overlay.test.tsx]: hey @matiasbavera, 
I've not been able to get a solution, hence I've removed the expect line for now. what are your thoughts?

Comment (matiasbavera) [reply packages/dashboard/src/components/schedule-visualizer/tests/robot-trajectories-overlay.test.tsx]: Ok, let's merge this PR, and can you create an issue on this? detailing what you want to achieve (trying to select an element inside a rendered component) and what is happening when it fails. Also what you tried to solve this issue.

Comment (matiasbavera) [reply packages/dashboard/src/components/schedule-visualizer/tests/robot-trajectories-overlay.test.tsx]: Maybe we can solve this by rendering some part of the component first, but I'm not sure. Let's discussed it in the issue.

**Review by matiasbavera — APPROVED** [review]
LGTM

File: packages/dashboard/src/components/schedule-visualizer/tests/robot-trajectories-overlay.test.tsx (hunk 13 above)
Comment (vallq) [reply packages/dashboard/src/components/schedule-visualizer/tests/robot-trajectories-overlay.test.tsx]: Alright! thanks @matiasbavera
